/ui: add `FontsLoader` component

saabi opened the pull request:
Closes #387

Also:
- fix `StorageIO` bug occurring when component is unmounted

mindrones commented on packages/docs/site/src/routes/_layout.svelte:
```diff
@@ -3,33 +3,45 @@
 	let showA11yMenu;
+
+	// $: console.log('typeface', $_a11ySettings.typeface.value);
+	// $: console.log('font families', $_fontFamilyNames);
```
leftovers

mindrones commented on packages/components/ui/src/drivers/fonts/README.md:
```diff
@@ -0,0 +1,68 @@
+`packages/docs/site/src/routes/_layout.svelte`.
+
+Props:
+- `FontsLoader`
```
I guess this line is unnecessary

mindrones commented on packages/components/ui/src/drivers/fonts/README.md:
```diff
@@ -0,0 +1,68 @@
+
+Props:
+- `FontsLoader`
+	- `families`: An array of objects describing the families and typefaces to load.
```
I think we need to explain:
- the format of the objects `{family, faces}`
- that not providing faces means assuming a system font (e.g. 'Courier New')

mindrones commented on packages/components/ui/src/drivers/fonts/FontsLoader.svelte:
```diff
@@ -0,0 +1,24 @@
+		$_fontFamilies = families;
+		$_status.isDone  = false;
+	};
+	$: isClientSide && ($_firstFamilyToLoad = firstFamilyToLoad);
+	$: isClientSide && loadFamilies($_runtimeFamilies);
```
Can we move these two within the `if` above?

saabi replied:
Unfortunately, moving them within the `if` clause will cause them to be executed more times than needed and load fonts from the network every time a new font is selected.

saabi replied:
Maybe we should add a comment explaining this behaviour.

saabi replied:
Since we're not using caching, these should remain as separate reactive statements for now.

mindrones commented on packages/components/ui/src/drivers/fonts/FontsLoader.svelte:
```diff
@@ -0,0 +1,24 @@
+<script>
+	import * as _ from 'lamb';
```
unused import

mindrones commented on packages/components/ui/src/drivers/fonts/FontsLoader.svelte:
```diff
@@ -0,0 +1,24 @@
+		$_fontFamilies = families;
+		$_status.isDone  = false;
+	};
+	$: isClientSide && ($_firstFamilyToLoad = firstFamilyToLoad);
```
For `$_fontFamilies = families;` and `$_firstFamilyToLoad = firstFamilyToLoad` I'm getting `no-unused-vars` so I'd prefer using `store.set(value)` instead

mindrones commented on packages/components/ui/src/drivers/fonts/stores.js:
```diff
@@ -0,0 +1,73 @@
+import * as _ from 'lamb';
+import {makeMergeAppliedFnMap} from '@svizzle/utils';
```
unused import

mindrones commented on packages/components/ui/src/io/storage/StorageIO.svelte:
```diff
@@ -40,7 +40,7 @@
 
 		return () => {
-			database.removeListener?.(key, setValue);
+			database.removeListener?.(key, updateStore);
```
[can't comment on line 24 as it's not part of the PR]

Re:

```
const updateStore = newValue => {
	// Current value in storage may be newer than value from event.
	newValue = database.getValue(key) || defaultValue;
	_store.set(newValue);
};
```

it's giving me a `no-param-reassign` eslint error, and I agree that reassigning an input var would be wrong, so I'd do something like

```
const updateStore = () => {
	// Current value in storage may be newer than value from event.
	const newValue = database.getValue(key) || defaultValue;
	_store.set(newValue);
};
```

as we don't seem to use `newValue` which seems to me an logical error?

mindrones commented on packages/components/ui/src/drivers/fonts/stores.js:
```diff
@@ -0,0 +1,73 @@
+			_.filterWith(_.not(isFirstToLoad(firstFamilyToLoad)))
+		]),
+	_.flatten,
+	getRuntimeFamilies
```
bad indentation

mindrones commented on packages/docs/site/src/routes/_layout.svelte:
```diff
@@ -3,33 +3,41 @@
+		defaultValue: fontFamilyNames,
+		values: fontFamilyNames,
+	}
+}}
```
bad indentation

mindrones commented on packages/components/ui/src/drivers/fonts/stores.js:
```diff
@@ -0,0 +1,73 @@
+const getRuntimeFamily = _.adapter([
+	_.casus(_.hasKey('faces'), _.pipe([
+		_.collect([
+			_.getKey('family'),
```
`_.getKey('family')` is defined multiple times: we need to define it once and reuse it

mindrones commented on packages/components/ui/src/drivers/fonts/stores.js:
```diff
@@ -0,0 +1,73 @@
+// deriveds
+export const _fontFamilyNames = derived(
+	_fontFamilies,
+	_.mapWith(_.getKey('family'))
```
(1) 

I would have a new file `packages/components/ui/src/drivers/fonts/utils.js` with:

```
export const getFamily = _.getKey('family');
export const getFamilyNames = _.mapWith(getFamily);
```

and any utility we use here, so that we would import `getFamilyNames` from `@svizzle/ui/src/drivers/fonts/utils` in the app.

mindrones commented on packages/docs/site/src/node_modules/app/config.js:
```diff
@@ -0,0 +1,76 @@
+	}
+];
+
+export const fontFamilyNames = fontFamilies.map(fontFamily => fontFamily.family);
```
I'd reuse `getFamilyNames` from `@svizzle/ui/src/drivers/fonts/utils`, see (1)

mindrones reviewed (changes requested):
First pass

mindrones commented on packages/components/ui/src/drivers/fonts/README.md:
```diff
@@ -0,0 +1,67 @@
+`packages/docs/site/src/routes/_layout.svelte`.
+
+Props:
+- `families`: An array of objects of shape `{family: string, faces: {src: string, descriptors: []}}` describing the families and typefaces to load. For information on the `descriptors` object please refer to https://developer.mozilla.org/en-US/docs/Web/API/FontFace/FontFace.
```
I think we still need to explain that not providing `faces` means assuming a system font (e.g. 'Courier New')

mindrones commented on packages/docs/site/src/node_modules/app/config.js:
```diff
@@ -0,0 +1,78 @@
+import {getFamilyNames} from "@svizzle/ui/src/drivers/fonts/utils";
```
We need to use single apices here

mindrones commented on packages/components/ui/src/drivers/fonts/utils.js:
```diff
@@ -0,0 +1,27 @@
+export const isFirstToLoad = firstFamilyToLoad => _.pipe([
+	getFamily,
+	_.is(firstFamilyToLoad)
+]);
```
I'd use: 

```js
export const isFamilyEqualTo = family => _.pipe([getFamily, _.is(family)]);
```

or

```js
export const isFamilyEqualTo = isKeyValue(['family', family]);
```

mindrones commented on packages/components/ui/src/drivers/fonts/README.md:
```diff
@@ -0,0 +1,67 @@
+
+Props:
+- `families`: An array of objects of shape `{family: string, faces: {src: string, descriptors: []}}` describing the families and typefaces to load. For information on the `descriptors` object please refer to https://developer.mozilla.org/en-US/docs/Web/API/FontFace/FontFace.
+- `firstFamilyToLoad`: If set, this property informs the component what family of typefaces to load first
```
We need to specify that it's a font family name, a string

mindrones commented on packages/components/ui/src/drivers/fonts/README.md:
```diff
@@ -0,0 +1,67 @@
+Props:
+- `families`: An array of objects of shape `{family: string, faces: {src: string, descriptors: []}}` describing the families and typefaces to load. For information on the `descriptors` object please refer to https://developer.mozilla.org/en-US/docs/Web/API/FontFace/FontFace.
+- `firstFamilyToLoad`: If set, this property informs the component what family of typefaces to load first
+- `status`: This property can be bound to for receiving the status of the font loading process.
```
We need to specify the properties of the object and their meaning

mindrones commented on packages/components/ui/src/drivers/fonts/stores.js:
```diff
@@ -0,0 +1,54 @@
+
+// main stores
+export const _firstFamilyToLoad = writable('');
+export const _fontFamilies = writable({});
```
this is an array

```suggestion
export const _fontFamilies = writable([]);
```

mindrones commented on packages/components/ui/src/drivers/fonts/FontsLoader.svelte:
```diff
@@ -0,0 +1,22 @@
+		loadFamilies
+	} from './stores';
+
+	export let families = {};
```
this is an array

```suggestion
	export let families = [];
```

mindrones commented on packages/components/ui/src/drivers/fonts/stores.js:
```diff
@@ -0,0 +1,54 @@
+	_fontFamilies,
+	getFamilyNames
+);
+export const _runtimeFamilies = derived(
+	[_fontFamilies, _firstFamilyToLoad],
+	([fontFamilies, firstFamilyToLoad]) => _.pipe([
+		_.collect([
+			_.filterWith(isFirstToLoad(firstFamilyToLoad)),
+			_.filterWith(_.not(isFirstToLoad(firstFamilyToLoad)))
+		]),
+		_.flatten,
+		getRuntimeFamilies
+	])(fontFamilies));
```
I think this:

```js
_.collect([
	_.filterWith(isFirstToLoad(firstFamilyToLoad)),
	_.filterWith(_.not(isFirstToLoad(firstFamilyToLoad)))
]),
_.flatten,
```

could be reduced to 

```js
_.sortWith([_.sorterDesc(isFirstToLoad(firstFamilyToLoad))])
```

But I'd prefer something like (2):
- note that arguments are sorted alphabetically
- I'm using `isFamilyEqualTo` as suggested elsewhere in the PR
- the utils that weer used would be unused with the code below so they'd need to be deleted from `utils.js`

(2)

```suggestion
export const _runtimeFamilies = derived(
	[_firstFamilyToLoad, _fontFamilies],
	([firstFamilyToLoad, fontFamilies]) => {
		const isFirstFamily = isFamilyEqualTo(firstFamilyToLoad);

		const runtimeFamilies = _.reduce(
			fontFamilies,
			(acc, {family, faces}) => {
				const fontFaces = faces.map(
					({src, descriptors}) => new FontFace(family, src, descriptors)
				);
				const isFirstFontFace = isFirstFamily(family);

				if (isFirstFontFace) {
					acc.unshift(fontFaces)
				} else {
					acc.push(fontFaces)
				}

				return acc;
			},
			[]
		);

		return runtimeFamilies;
	}
);
```

mindrones reviewed (changes requested):
pass two

mindrones reviewed (approved):
Great job! 👌

saabi committed:
/ui: add `FontsLoader` component

Closes #387

Also:
- fix `StorageIO` bug occurring when component is unmounted
- removed duplicate `isServerSide` definition
